File: Design-Patterns-Using-Cpp/Adapter/AdapterPatternRectangle.cpp
#include <iostream>

// Desired Interface
class Shape {
public:
    virtual void draw() = 0;
    virtual ~Shape() {} // Virtual destructor for proper cleanup
};

// Circle class
class Circle : public Shape {
private:
    int x, y, radius;

public:
    Circle(int x, int y, int r) : x(x), y(y), radius(r) {}

    void draw() override {
        std::cout << "Draw the Circle." << std::endl;
    }
};

// Rectangle class (Adaptee)
class Rectangle {
private:
    int x, y, length, width;

public:
    Rectangle(int x, int y, int l, int w) : x(x), y(y), length(l), width(w) {}

    void oldDraw() {
        std::cout << "Drawing Rectangle." << std::endl;
    }
};

// RectangleAdapter class
class RectangleAdapter : public Shape {
private:
    Rectangle adaptee;

public:
    RectangleAdapter(int x, int y, int l, int w) : adaptee(x, y, l, w) {}

    void draw() override {
        adaptee.oldDraw();
    }
};

// Client Code
int main() {
    Shape* adapter = new RectangleAdapter(1, 2, 3, 4);
    adapter->draw();

    // Clean up memory
    delete adapter;

    return 0;
}
/*
Drawing Rectangle.
*/

File: Design-Patterns-Using-Swift/Adapter/AdapterPatternRectangle.swift
// Desired Interface
protocol Shape {
    func draw()
}

// Circle class
class Circle: Shape {
    private var x: Int
    private var y: Int
    private var radius: Int

    init(x: Int, y: Int, radius: Int) {
        self.x = x
        self.y = y
        self.radius = radius
    }

    func draw() {
        print("Draw the Circle.")
    }
}

// Rectangle class (Adaptee)
class Rectangle {
    private var x: Int
    private var y: Int
    private var length: Int
    private var width: Int

    init(x: Int, y: Int, length: Int, width: Int) {
        self.x = x
        self.y = y
        self.length = length
        self.width = width
    }

    func oldDraw() {
        print("Drawing Rectangle.")
    }
}

// RectangleAdapter class
class RectangleAdapter: Shape {
    private var adaptee: Rectangle

    init(x: Int, y: Int, length: Int, width: Int) {
        self.adaptee = Rectangle(x: x, y: y, length: length, width: width)
    }

    func draw() {
        adaptee.oldDraw()
    }
}

// Client Code
let adapter: Shape = RectangleAdapter(x: 1, y: 2, length: 3, width: 4)
adapter.draw()

/*
Drawing Rectangle.
*/
